tests/test_kgain_cal.py

```diff
@@ -0,0 +1,847 @@
+        if arr[i] == arr[i - 1]:
+            current_count += 1
+        else:
+            counts.append(current_count)
+            current_count = 1
+    counts.append(current_count)  # append the count of the last element
+    return counts
+
+############### make stacks with simulated frames ##################
+
+# path to nonlin table made from running calibrate_nonlin.py on TVAC frames
+# table used only to choose parameters to make analytic nonlin functions
+here = os.path.abspath(os.path.dirname(__file__))
+nonlin_table_path = Path(here,'test_data','nonlin_table_TVAC.txt')
+nonlin_flag = False # True adds nonlinearity to simulated frames
+
+# Load the arrays needed for calibrate_nonlin function from the .npz file
+loaded = np.load(Path(here,'test_data','nonlin_arrays_ut.npz'))
+# Access the arrays needed for calibrate_nonlin function
+exp_time_stack_arr0 = loaded['array1']
+len_list0 = loaded['array3']
+
+# Load the flux map
+hdul =  fits.open(Path(here,'test_data','FluxMap1024.fits'))
+fluxmap_init = hdul[0].data
+hdul.close()
+fluxmap_init[fluxmap_init < 50] = 0 # cleanup flux map a bit
+fluxMap = 0.8*fluxmap_init # e/s/px, for G = 1
+
+# assumed detector parameters
+kgain_in = 8.7 # e-/DN
+rn_in = 130 # read noise in e-
+bias = 2000 # e-
+emgain = 1.0
+
+# cubic function nonlinearity for emgain of 1
+if nonlin_flag:
+    coeffs_1, DNs, _ = nonlin_coefs(nonlin_table_path,1.0,3)
+else:
+    coeffs_1 = [0.0, 0.0, 0.0, 1.0]
+    _, DNs, _ = nonlin_coefs(nonlin_table_path,1.0,3)
+
+frame_list2 = []
+# make 30 uniform frames with emgain = 1
+## DATA: make_frame
+for j in range(30):
+    frame2 = make_frame(fluxMap,bias,kgain_in,rn_in,emgain,7.0,coeffs_1,nonlin_flag)
+    frame_list2.append(frame2)
+# Remove singleton dimensions
+stack_arr2 = np.squeeze(np.stack(frame_list2))
+
+index = 0
+iG = 0 # doing only the em gain = 1 case
+g = emgain
+exp_time_loop = exp_time_stack_arr0[index:index+len_list0[iG]]
+index = index + len_list0[iG]
+frame_list = [] # initialize frame stack
+if nonlin_flag:
+    coeffs, _, vals = nonlin_coefs(nonlin_table_path,g,3)
+else:
+    coeffs = [0.0, 0.0, 0.0, 1.0]
+    vals = np.ones(len(DNs))
+
+stack_list = []
+exp_repeat_counts = count_contiguous_repeats(exp_time_loop)
+for j in range(len(exp_repeat_counts)):
+    frame_list = []
+    for t in range(exp_repeat_counts[j]):
+        # Simulate full frame
+        exp_time = exp_time_loop[t+j*exp_repeat_counts[j]]
+        frame_sim = make_frame(fluxMap,bias,kgain_in,rn_in,g,
+                               exp_time,coeffs,nonlin_flag)
+        frame_list.append(frame_sim)
+    frame_stack = np.stack(frame_list)
+    stack_list.append(frame_stack)
+# Remove singleton dimensions 
+stack_arr = np.squeeze(np.stack(stack_list))
+
+# set input parameters for calibrate_kgain function
+min_val = 800
+max_val = 3000
+binwidth = 68
+
+################### define class for tests ######################
+
+class TestCalibrateKgain(unittest.TestCase):
```

Comment:
It would be good to reformat these tests to be a bit more readable and make it clearer where the main functionality is being tested (see some of the other existing tests for examples.

Reply:
Done